lift/Readme.md

```diff
@@ -6,271 +6,360 @@
 ```
 
-This means that every migration will have its own generated Prisma Client in the chosen language (JS, TS or Go).
+The order of execution is the following:
 
-### Writing `before` or `after` migration scripts
+1. `20190920142120-add-user/after.sh`
+2. `20190920142120-add-user/steps.json`
+3. `20190920142120-add-user/before.sh`
+4. `20190920142118-initial/after.sh`
+5. `20190920142118-initial/steps.json`
+6. `20190920142118-initial/before.sh`
 
-There are a few use-cases when you want to run your own custom migration logic:
+### Lift Server
 
-- Making sure that the data adheres to constraints that you want to introduce, like `unique` or `non null` constraints
-- Adding specific database primitives like functions in Postgres, which are not yet supported by Prisma
-- Initializing data after a migration has been executed. Both SQL or the Prisma Client could be useful here
+The Lift server is a low-level interface that the Lift Client communicates with. Currently the client communicates to the migration server in the JSONRPC format
+over stdio. In the future, we'll migrate this to REST or JSONRPC over HTTP.
 
-Let's say that after Prisma has migrated the database to the datamodel defined in `migrate/20190322092247-my-initial-migration/`, you want to insert some seed
-data. The way you can do this, is by either defining a shell script or an executable, either called `before.EXT`, while EXT is your favorite file extension as
-`.sh` or `.js` or `after.EXT`. A file just called `before` or `after` is also valid.
+**Note** This API is subject to change
 
-Which file will be executed when is decided by a convention. When the Prisma CLI opens a migration folder, it will perform the following actions:
+#### inferMigrationSteps
 
-1. Check if there is only maximum one file with the `after` prefix and only maximum one file with the `before` prefix.
-2. Check for the existence of a before file and try to execute it. The following filenames are valid:
+Calculates the Steps needed to transition the datasources from the current state to the next state. This is called by the Save method in the Lift Client.
 
-| Filename    | Action         |
-| ----------- | -------------- |
-| before.sh   | sh before.sh   |
-| before.bash | bash before.sh |
-| before      | ./before       |
-| before.js   | ./before.js    |
-| before.ts   | ./before.ts    |
+#### applyMigration
 
-3. Migrate the datamodel
-4. Check for the existence of an after file and try to execute it. The following filenames are valid:
+Applies the Steps we inferred. This is called by the Up method in the Lift Client.
 
-| Filename   | Action        |
-| ---------- | ------------- |
-| after.sh   | sh after.sh   |
-| after.bash | bash after.sh |
-| after      | ./after       |
-| after.js   | ./after.js    |
-| after.ts   | ./after.ts    |
+#### unapplyMigration
 
-### Up & Down in migration scripts
+Unapplies the Steps in the migrations folder. This is called by the Down method in the Lift Client.
```

Comment:
This is not entirely correct. `unapplyMigration` just rolls back the last migration in the database. This method is called repeatedly by the CLI.

Reply:
good call